readme.md
# EDD Bookings

[![Code Climate](https://codeclimate.com/github/RebelCodecom/edd-bookings/badges/gpa.svg)](https://codeclimate.com/github/RebelCodecom/edd-bookings)

Extends the Easy Digital Downloads WordPress plugin with booking capabilities.

## Description

This extension adds a booking system to WordPress, allowing customers to book and pay for appointments, meetings, consultations and other scheduled events, service or resources that need to be scheduled by date and time against a payment.

Site administrators are provided with a powerful interface that provides complete control over dates and prices for bookings. Once configured, customers are presented with an attractive and intuitive calendar interface that shows all available time slots. To book a date and time, customers simply need to choose the date on the calendar and then proceed with the purchase.

All bookings are presented within the WordPress dashboard, giving the site administrator easy access at a glance to all the incoming bookings and payments.

### Feature Highlights

* Simple, intuitive configuration for dates and times
* Calendar builder for configuring dates and times available to be booked
* Supports single and multiple sessions per booking
* Works with all payment gateways
* Manage bookings from the WordPress admin area
* Fully integrated with Frontend Submissions

### Integrations

#### EDD Frontend Submissions

EDD Bookings is fully integrated with the Frontend Submissions extension, allowing you to operate multi-vendor marketplaces!

Vendors can submit booking products with the same controls as the site administrator. From the vendor dashboard, bookings are displayed to vendors with all of the necessary appointment information. Vendors even have their own calendar view.

If you are using the Commissions extension, you may also send vendors information about a new booking using the `{booking}` email tag.

## Requirements:

* **PHP**: 5.3+
* **WordPress**: 4.0+
* **Easy Digital Downloads**: 2.4.0+

Website: www.eddbookings.com

License: [GPLv3](https://www.gnu.org/licenses/gpl-3.0.en.html)

## Building the Plugin Archive

This process requires PHP 5.6 or later.

Obtain developer tools with

    composer install

Invoke the Phing `release` target with the `release=X.Y.Z` argument:

    vendor/bin/phing release -Dversion=x.y.z

The release files can be found in `build/` and the archive for uploading and installing into WordPress can be found in `releases/`. The build process will automatically include all non-dev dependencies.


changelog.txt
2.2.5 (2017-10-18)
- Switched licensing server; licenses and updates are now handled by RebelCode.

2.2.4 (2017-07-25)
- Fixed: Checkout fatal error and duplicate purchase buttons with EDD v2.8.
- Fixed: Fatal error on saving a Download with translated time units.
- Fixed: Multiple strings were not translatable.
- Fixed: FES fields were required even when bookings were disabled.
- Fixed: Fatal error if the `phpxml` extension was not installed.

2.2.3 (2017-05-19)
- Fixed: The calendar was not formatting the price correctly.

2.2.2 (2017-04-24)
- Fixed: Admin bookings calendar popup hides when the page is scrolled.
- Fixed: FES Vendor bookings calendar popup did not show.
- Fixed: Duration picker did not allow choosing more than 1 session for day and week session units.

2.2.1 (2017-04-12)
- Improved: Calendar was taking a long time to load on the front-end.
- Fixed: Incorrect start time when using a session length that was not a multiple of 10.
- Fixed: The minimum and maximum number of allowed sessions now swap when applicable.
- Fixed: The session length and minimum and maximum number of sessions default to 1 if less than zero.
- Fixed: FES field options are now required on the submissions form.
- Fixed: The bookings enabled checkbox correctly saves from the FES submissions form.

2.2.0 (2016-12-14)
- New: Added the ability to create and edit bookings from WP Admin.
- New: Added an availability preview when creating or editing bookable Downloads.
- Improved: The checkout error message is now customizable through an admin option.
- Improved: Booking end times are now more user friendly.
- Fixed: Time rules for days of the week were generating incorrect times for certain timezones.
- Fixed: The front-end calendar was not being shown when using the `purchase_link` shortcode.
- Fixed: The availability builder was not clickable on Firefox.
- Fixed: Booking data was not being saved for Downloads with bookings disabled.
- Fixed: Autoloader not found when plugin is used in a Composer-managed WordPress installation.

2.1.3 (2016-11-08)
- Improved client-side script logic and performance.
- Improved compatability with other EDD extensions that add items to the cart.
- Improved checkout - bookable downloads' sessions can be edited before checkout.
- Improved checkout - a message is shown for items in the cart without a booked session.
- Improved some translation strings.
- Improved translations for client-side scripts.
- Fixed a bug that caused the first available day to be offset with the session length.
- Fixed a bug that caused Downloads to have incomplete meta data default values.

2.1.2 (2016-10-07)
- Added an global option for FES that controls the default single-page display of bookable Downloads.
- Fixed a bug that caused the FES field to disallow submission when set to "required".
- Fixed strings and text domain problems in source code for translation purposes.
- Fixed error when activating EDD Bookings when EDD is deactivated.
- Fixed JS error when frontend calendar fails to retrieve information from the server.
- Fixed corrupted Download meta data when exporting/importing from the WordPress importer.

2.1.1 (2016-09-16)
- Fixed prices for bookable downloads.
- Fixed error on Submissions Form on linux/unix systems.
- Fixed the availability builder on the FES submissions page when the FES CSS option was disabled.
- Fixed a bug where the calendars would not appear on the site when Yoast SEO was installed.
- Fixed various visual problems with the calendars on most themes.
- Fixed broken image link in the "What's New?" page.

2.1.0 (2016-08-23)
- Integration with EDD Frontend Submissions extension.

2.0.1 (2016-07-08)
- Improved Download options. Tooltips and labels are now clearer.
- The multi-view option is now inverted, making it more intuitive to use.
- Added a notice for Downloads without availability times.
- Fixed broken documentation link.
- Fixed calendar infinite loading bug on Decembers.
- Fixed time fields in availability builder not allowing manual input.
- Fixed current time not using the site timezone under the availability builder.
- Fixed broken "%s" placeholders on the frontend.

2.0.0 (2016-06-22)
- The plugin is now powered by a new bookings library under the hood.
- Improved the frontend calendar, fixing various bugs and performance issues.
- Reworked the admin interface, making it simpler and cleaner.
- Added the new Admin Calendar in the admin interface for easier booking management.
- Improved handling of timezone difference between your site and your customers.

1.0.3 (2016-02-09)
- Added integration with the EDD `[purchase_link]` shortcode.
- Added the `booking_options` option to the `[purchase_link]` shortcode.
- Fixed a bug where changing the value of the duration field on the front-end caused a JavaScript error that cancels validation.
- Improved front-end JavaScript so that it is compatible with themes that do not use `post_class`.

1.0.2 (2016-01-21)
- Added support for server and local timezones.
- Added booking information to the purchase confirmation receipt.
- Fixed a bug where certain booked sessions where not removed from the availability.
- Fixed bug where viewing a Bookings page created a revision for that booking.
- Fixed a bug where an availability with only time or time-group ranges did not auto generate day ranges.
- Fixed various script and style bugs when viewing bookable downloads in pages that contain multiple bookable downloads.
- Fixed a bug where error messages where being displayed on the front-end by default.

1.0.1 (2015-12-21)
- Fixed prices being set to zero when extension is active.
- Fixed bug where orders cannot be updated.
- Fixed bug where downloads with Custom Ranges sometimes could not be saved.

1.0.0 (2015-11-03)
- First stable release.


edd-bookings.php
<?php

/**
 * @wordpress-plugin
 * Plugin Name: Easy Digital Downloads - Bookings
 * Plugin URI: https://easydigitaldownloads.com/downloads/edd-bookings/
 * Description: Adds a simple booking system to Easy Digital Downloads.
 * Version: 2.2.5
 * Author: RebelCode
 * Text Domain: eddbk
 * Domain Path: /languages/
 * License: GPLv3
 */

/**
 * Copyright (C) 2015-2017 RebelCode Ltd.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// If the file is called directly, or has already been called, abort
if (!defined('WPINC') || defined('EDD_BK')) {
    die;
}

// Plugin File Constant
define('EDD_BK', __FILE__);
// Plugin Version
define('EDD_BK_VERSION', '2.2.5');
// Plugin ID, or slug-esque name
define('EDD_BK_PLUGIN_ID', 'eddbk');
// Plugin Name
define('EDD_BK_PLUGIN_NAME', 'EDD Bookings');
// Plugin Author
define('EDD_BK_PLUGIN_AUTHOR', 'RebelCode');
// Parent Plugin Class name
define('EDD_BK_PARENT_PLUGIN_CLASS', 'Easy_Digital_Downloads');
// Parent Plugin Min version required
define('EDD_BK_PARENT_PLUGIN_MIN_VERSION', '2.4.0');
// Minimum WordPress version
define('EDD_BK_MIN_WP_VERSION', '4.0');
// Minimum PHP Version
define('EDD_BK_MIN_PHP_VERSION', '5.3');
// Database version number
define('EDD_BK_DB_VERSION', '1');
// Default text domain
define('EDD_BK_TEXT_DOMAIN', 'eddbk');

// Plugin Name on the server that provides the updates
define('EDD_BK_PLUGIN_NAME_SERVER', EDD_BK_PLUGIN_NAME);
// URL to the server that provides the updates
define('EDD_BK_UPDATES_SERVER_URL', 'https://eddbookings.com/edd-sl/');

// Documentation link
define('EDD_BK_DOCS_URL', 'http://docs.easydigitaldownloads.com/category/1100-bookings');

// Initialize Directories
define('EDD_BK_DIR', plugin_dir_path(EDD_BK));
define('EDD_BK_BASE', plugin_basename(EDD_BK));
define('EDD_BK_VENDOR_DIR', EDD_BK_DIR . 'vendor/');
define('EDD_BK_AUTOLOAD_FILE', EDD_BK_VENDOR_DIR . 'autoload.php');
define('EDD_BK_LANG_DIR', dirname(EDD_BK_BASE) . '/languages');
define('EDD_BK_CONFIG_DIR', EDD_BK_DIR . 'config/');
define('EDD_BK_INCLUDES_DIR', EDD_BK_DIR . 'includes/');
define('EDD_BK_VIEWS_DIR', EDD_BK_DIR . 'views/');
define('EDD_BK_ASSETS_DIR', EDD_BK_DIR . 'assets/');
define('EDD_BK_JS_DIR', EDD_BK_ASSETS_DIR . 'js/');
define('EDD_BK_JS_LIB_DIR', EDD_BK_JS_DIR . 'lib/');
define('EDD_BK_ADMIN_DIR', EDD_BK_INCLUDES_DIR . 'admin/');
define('EDD_BK_PUBLIC_DIR', EDD_BK_INCLUDES_DIR . 'public/');
define('EDD_BK_DOWNLOADS_DIR', EDD_BK_INCLUDES_DIR . 'downloads/');
define('EDD_BK_BOOKINGS_DIR', EDD_BK_INCLUDES_DIR . 'bookings/');
define('EDD_BK_CUSTOMERS_DIR', EDD_BK_INCLUDES_DIR . 'customers/');
define('EDD_BK_EXCEPTIONS_DIR', EDD_BK_INCLUDES_DIR . 'exceptions/');
define('EDD_BK_WP_HELPERS_DIR', EDD_BK_INCLUDES_DIR . 'wp-helpers/');
define('EDD_BK_INTEGRATIONS_DIR', EDD_BK_DIR . 'integrations/');

// Set up the uploads directory
$uploadsDir = wp_upload_dir();
$eddBookingsUploadDir = array($uploadsDir['basedir'], 'edd-bookings');
$eddBookingsUploadUrl = array($uploadsDir['baseurl'], 'edd-bookings');
define('EDD_BK_UPLOADS_DIR', implode(DIRECTORY_SEPARATOR, $eddBookingsUploadDir));
define('EDD_BK_UPLOADS_URL', implode('/', $eddBookingsUploadUrl));
if (!file_exists(EDD_BK_UPLOADS_DIR)) {
    wp_mkdir_p(EDD_BK_UPLOADS_DIR);
}

// Initialize URLs
define('EDD_BK_PLUGIN_URL', plugin_dir_url(EDD_BK));
define('EDD_BK_ASSETS_URL', EDD_BK_PLUGIN_URL . 'assets/');
define('EDD_BK_CSS_URL', EDD_BK_ASSETS_URL . 'css/');
define('EDD_BK_JS_URL', EDD_BK_ASSETS_URL . 'js/');
define('EDD_BK_IMGS_URL', EDD_BK_ASSETS_URL . 'imgs/');
define('EDD_BK_FONTS_URL', EDD_BK_ASSETS_URL . 'fonts/');
define('EDD_BK_INTEGRATIONS_URL', EDD_BK_PLUGIN_URL . 'integrations/');

// For Debugging
define('EDD_BK_DEBUG', FALSE);

// Check minimum php version
if (version_compare(PHP_VERSION, EDD_BK_MIN_PHP_VERSION, '<')) {
    // load plugins.php file from WordPress if not loaded
    require_once(ABSPATH . 'wp-admin/includes/plugin.php');
    deactivate_plugins(__FILE__);
    wp_die(
        sprintf(
            _x('EDD Bookings requires PHP version %s or later.', 'Example: EDD Bookings requires PHP 5.3 or later', 'eddbk'),
            EDD_BK_MIN_PHP_VERSION
        ),
        __('EDD Bookings has been disabled', 'eddbk'),
        array(
            'back_link' => true
        )
    );
}

// Check if vendor dir and autoload file exist
if (is_dir(EDD_BK_VENDOR_DIR) && file_exists(EDD_BK_AUTOLOAD_FILE)) {
    // Load the autoloader file
    require EDD_BK_AUTOLOAD_FILE;
}
// This message should never be displayed to users. It only exists for the sake of development as a reminder that dependencies need to be installed.
else if (!class_exists('Composer\Autoload\ClassLoader')) {
    // load plugins.php file from WordPress if not loaded
    require_once(ABSPATH . 'wp-admin/includes/plugin.php');
    deactivate_plugins(__FILE__);
    wp_die(
        sprintf(
            '%1$s<br/>%2$s <code>composer install</code><br/>%3$s',
            __('Plugin dependencies are not loaded!', 'eddbk'),
            __('If you are using Composer with WordPress, make sure that the autoload.php file being loaded.', 'eddbk'),
            __('The EDD Bookings plugin will be deactivated.', 'eddbk')
        ),
        __('Plugin dependencies are not installed!', 'eddbk'),
        array(
            'back_link' => true
        )
    );
}

/**
 * Gets the plugin main class singleton instance.
 * 
 * @staticvar Aventura\Edd\Bookings\Plugin $instance The singleton instance
 * @return \Aventura\Edd\Bookings\Plugin The singleton instance.
 */
function eddBookings()
{
    /* @var $instance \Aventura\Edd\Bookings\Plugin */
    static $instance = null;
    // If null, instantiate
    if (is_null($instance)) {
        // Create the factory
        $defaultFactoryClass = 'Aventura\\Edd\\Bookings\\Factory';
        $filteredFactoryClass = apply_filters('edd_bk_main_factory_class', $defaultFactoryClass);
        $factoryClass = (is_subclass_of($filteredFactoryClass, $defaultFactoryClass))
                ? $filteredFactoryClass
                : $defaultFactoryClass;
        /* @var $factory \Aventura\Edd\Bookings\Factory */
        $factory = new $factoryClass();
        // Create the plugin
        $instance = $factory->create();
        // Set factory's parent plugin pointer
        $factory->setPlugin($instance);
        // Throw exception if the filtered factory class did not exist and the default had to be used
        if ($filteredFactoryClass !== $defaultFactoryClass && $factoryClass === $defaultFactoryClass) {
            $msg = sprintf('The %s class does not exist or is not a valid factory class. The default was used.', $filteredFactoryClass);
            throw new Exception($msg);
        }
    }
    return $instance;
}

// Activation/Deactivation hooks
register_activation_hook(__FILE__, array(eddBookings(), 'onActivate'));
register_deactivation_hook(__FILE__, array(eddBookings(), 'onDeactivate'));

eddBookings()->getI18n()->loadTextDomain();

// Load the integrations
eddBookings()->loadDirectory('integrations');

// Hook in the plugin - In actuality, the plugin is registering its hooks with the Hook Manager
eddBookings()->hook();
// This makes the Hook Manager register the saved hooks to WordPress
eddBookings()->getHookManager()->registerHooks();
